Fixes #RHIROS-1231, #RHIROS-1172, #RHIROS-1175, #RHIROS-1177 - store suggested instance type info and api changes

upadhyeammit opened the pull request:
## Why do we need this change? :thought_balloon:

This pull request adds below changes,

1. Store suggested instance types and its price in separate columns
2. Api implementation of SuggestedInstanceTypes
3. Filtering logic added for api endpoint
4. Sorting logic added for api endpoint

## Documentation update? :memo:

- [ ] Yes
- [x] No

## Security Checklist :lock:

Upon raising this PR please go through [RedHatInsights/secure-coding-checklist](https://github.com/RedHatInsights/secure-coding-checklist)

## :guardsman: Checklist :dart:

- [ ] Bugfix
- [x] New Feature
- [ ] Refactor
- [ ] Unittests Added
- [ ] DRY code
- [ ] Dependency Added
- [x] DB Migration Added

## Additional :mega:

Consideration before making decision to add extra columns,

1. Completely normalize the rule_hit_details column to separate tables as instance_type and instance_price are subset of data stored in rule_hit_details column. Though this makes difference however there should be separate effort for that, as it needs code changes in many of the project structure.
5. Have a separate table for suggested_instance_type at least. This makes sense however as of now there only two data points effectively will be used for api which are instance_type and instance_price. Apart from that api suppose to return description which is static in nature and should be built using constants, and provider is only aws now.
6. The instance_price is not required for suggested_instance api however that can simplify the executive report.
7. There are some irrelevant linting changes in this pr, that is to satisfy the flake8

upadhyeammit commented:
To check how much time it an take to update existing rows in performance_profile table I did below steps. I took 10 million rows as example, and I can see it takes lot of time(5.38 hours). I am yet to check whats the size of the table in production however below are the details,

~~~
# create table performance_profile_copy(
like performance_profile
including indexes);

## Need to drop this as random num for system_id are not unique
# alter table performance_profile_copy drop constraint performance_profile_copy_pkey;

## Dropping this index as in migration index is getting created after updating the rows
# drop index performance_profile_copy_instance_type_idx;

# There is slight change in rule_hit_details, here is just one rule and not array of rules, this is to keep it simple

# insert into performance_profile_copy(performance_record, performance_utilization, report_date, system_id, number_of_recommendations, state, operating_system, rule_hit_details, psi_enabled) select '{"hinv.ncpu": 2.0}', '{"hinv.ncpu": 2.0}', now() + round(random()*1000) * '1 second'::interval, round(random()*10000), round(random()*4),(ARRAY['Oversized', 'Undersized', 'Idling', 'Under pressure','Storage rightsizing','Optimized','Waiting for data'])[round(random())+7], '{"hinv.ncpu": 2.0}', '{"rule_id": "ros_instance_evaluation|INSTANCE_IDLE", "component": "telemetry.rules.plugins.ros.ros_instance_evaluation.report", "type": "rule", "key": "INSTANCE_IDLE", "details": {"rhel": "8.4", "cloud_provider": "aws", "instance_type": "t2.micro", "region": "us-east-1", "price": 0.0116, "candidates": [["t2.nano", 0.0058]], "states": {"idle": ["IDLE"]}, "type": "rule", "error_key": "INSTANCE_IDLE"}, "tags": [], "links": {"jira": ["https://issues.redhat.com/browse/CEECBA-5875"], "kcs": []}, "system_id": "a319580a-2321-47c7-b061-886548bea067"}', (ARRAY[false,true])[round(random())+1] from generate_series(1,10000000);

# \timing on
Timing is on.

# update performance_profile_copy set instance_type = subquery.instance_type, instance_price = cast(subquery.instance_price as double precision) from (select rule_hit_details['details']['candidates'][0][0] as instance_type, rule_hit_details['details']['candidates'][0][1] as instance_price, system_id from performance_profile_copy) as subquery where performance_profile_copy.system_id = subquery.system_id;

Time: 19397110.988 ms (05:23:17.111)
~~~

upadhyeammit commented:
I can see prod db has only 1614 systems, so considering 1700 system counts I can see it takes 23 miliseconds only, so this should work for us,
~~~
postgres=# insert into performance_profile_copy(performance_record, performance_utilization, report_date, system_id, number_of_recommendations, state, operating_system, rule_hit_details, psi_enabled) select '{"hinv.ncpu": 2.0}', '{"hinv.ncpu": 2.0}', now() + round(random()*1000) * '1 second'::interval, round(random()*10000), round(random()*4),(ARRAY['Oversized', 'Undersized', 'Idling', 'Under pressure','Storage rightsizing','Optimized','Waiting for data'])[round(random())+7], '{"hinv.ncpu": 2.0}', '{"rule_id": "ros_instance_evaluation|INSTANCE_IDLE", "component": "telemetry.rules.plugins.ros.ros_instance_evaluation.report", "type": "rule", "key": "INSTANCE_IDLE", "details": {"rhel": "8.4", "cloud_provider": "aws", "instance_type": "t2.micro", "region": "us-east-1", "price": 0.0116, "candidates": [["t2.nano", 0.0058]], "states": {"idle": ["IDLE"]}, "type": "rule", "error_key": "INSTANCE_IDLE"}, "tags": [], "links": {"jira": ["https://issues.redhat.com/browse/CEECBA-5875"], "kcs": []}, "system_id": "a319580a-2321-47c7-b061-886548bea067"}', (ARRAY[false,true])[round(random())+1] from generate_series(1,1700);
INSERT 0 1700
Time: 8.150 ms
postgres=# select count(system_id) from performance_profile_copy;
-[ RECORD 1 ]
count | 1700

Time: 0.639 ms
postgres=# update performance_profile_copy set instance_type = subquery.instance_type, instance_price = cast(subquery.instance_price as double precision) from (select rule_hit_details['details']['candidates'][0][0] as instance_type, rule_hit_details['details']['candidates'][0][1] as instance_price, system_id from performance_profile_copy) as subquery where performance_profile_copy.system_id = subquery.system_id;
UPDATE 1700
Time: 24.784 ms
~~~

upadhyeammit commented:
/retest

upadhyeammit commented:
I am able to reproduce this! Its happening because on production I can see pgsql version 13.7 and the local uses the latest which is 15.x. On the latest version same update query works, I am checking how to make query compliant to 13.7 version!

upadhyeammit commented:
> I am able to reproduce this! Its happening because on production I can see pgsql version 13.7 and the local uses the latest which is 15.x. On the latest version same update query works, I am checking how to make query compliant to 13.7 version!

This is ready for review now with latest changes!

r14chandra commented on ros/processor/insights_engine_consumer.py:
```diff
@@ -199,7 +210,10 @@ def process_report(self, host, platform_metadata, reports, system_metadata, perf
                     "operating_system": system.operating_system,
                     'psi_enabled': system_metadata.get('psi_enabled'),
+                    'instance_type': suggested_instance_type,
+                    'instance_price': suggested_instance_price,
```
Is it really required to separate out an instance type and its price? IMO there should be one column that would contain an array list of all the candidates and their respective prices.
ex: `candidates`: `[ [type1, price1], [type2, price2], ...]` 
Then we can easily get the first pair from the array as per our requirement. 
This arises as we also show the first three candidates in the recommendation section. 

Let me know your thoughts on this. Thanks!

kgaikwad reviewed (changes requested):
Thank you @upadhyeammit for pull-request.
Added few inline suggestions. My main concern is with naming. One might get confuse with actual system's instance_type. 

Also, could you update commit message as this will be parent commit? Example - `Fixes #RHIROS-1175 - store suggested instance type info into columns`

upadhyeammit commented on ros/lib/utils.py:
```diff
@@ -151,18 +151,35 @@ def insert_performance_profiles(session, system_id, fields):
 
-    for model_class in [PerformanceProfile, PerformanceProfileHistory]:
-        new_entry = model_class(**fields)
+    model_fields_map = {PerformanceProfile: fields,
+                        PerformanceProfileHistory: fields_for_perf_profile_history}
+
+    for model_class, model_fields in model_fields_map.items():
+        new_entry = model_class(**model_fields)
```
Actually to simplify this if else I thought to use static dict, like to know if I can keep it?

upadhyeammit commented on ros/processor/insights_engine_consumer.py:
```diff
@@ -34,6 +34,15 @@
 producer = None
 
 
+def get_instance_and_price_from_rule_hit(rule_hit_details, state):
```
I did not do it because `state_without_instance_types` is not being used anywhere else in code? and so I kept the variable local to method as well.

upadhyeammit commented on ros/processor/insights_engine_consumer.py:
```diff
@@ -34,6 +34,15 @@
 producer = None
 
 
+def get_instance_and_price_from_rule_hit(rule_hit_details, state):
+    instance_and_price = [None, None]
+    state_without_instance_types = ["OPTIMIZED", "NO_PCP_DATA"]
+    if state not in state_without_instance_types:
+        all_instances = rule_hit_details[0]['details']['candidates']
+        instance_and_price = all_instances[0]
+    return instance_and_price
```
I kept it as local variable because I can't see its being shared with anything else in code? and I agree to changing the name, and refactoring!

kgaikwad commented on ros/lib/utils.py:
```diff
@@ -151,18 +151,37 @@ def insert_performance_profiles(session, system_id, fields):
+        fields_for_perf_profile_history = fields.copy()
+        del fields_for_perf_profile_history['top_candidate']
+        del fields_for_perf_profile_history['top_candidate_price']
+    except KeyError:
+        LOG.debug("Key(s) top_candidate and/or top_candidate_price not found!")
```
```suggestion
    except KeyError as err:
        LOG.debug(f'Key(s) not found in fields: {error}')
```

Can we get display key that actually not present using ^^ ?

upadhyeammit commented:
> Observation: attached ticket is to add endpoint. If the API endpoint work is in-progress and not ready, I would suggest to create a separate card for DB changes and attach here instead of RHIROS-1175. In that way, both changes can go separately.

yeah, api changes are work in progress, I am creating new card for db changes, would update the pr and commit

upadhyeammit commented:
I have pushed api commit to this pr, I just felt either we should merge this with one commit which is of db migration or if we are not merging then I can keep pushing to this pr as changes are dependent

upadhyeammit commented:
/retest

upadhyeammit commented:
/retest

kgaikwad commented on ros/api/v1/suggested_instance_types.py:
```diff
@@ -0,0 +1,84 @@
+            PerformanceProfile.top_candidate)
+
+    @property
+    def instance_descriptions(self):
+        instance_and_descriptions = {}
+        for instance, info in INSTANCE_TYPES.items():
+            processor = info['extra']['physicalProcessor']
+            v_cpu = info['extra']['vcpu']
+            memory = info['extra']['memory']
+            instance_and_descriptions[instance] = f"{processor} instance with {v_cpu} vCPUs and {memory} RAM"
+        return instance_and_descriptions
```
Suggestion:
Can we move this method to lib/utils or somewhere else? 
Also, it looks like these description strings will rarely change. Can we implement caching around this method and set an expiry to re-generate this dict?

upadhyeammit replied:
Moved to separate file, looking further on adding redis cache

upadhyeammit replied:
Cached this to infinite time(timeout=0). This means when our application will restart on deployment the cache will be reloaded.

kgaikwad commented on ros/lib/models.py:
```diff
@@ -26,6 +28,7 @@ class PerformanceProfile(db.Model):
             ondelete='CASCADE'),
         db.Index('non_optimized_system_profiles', number_of_recommendations, unique=False,
                  postgresql_where=(number_of_recommendations > 0)),
+        db.Index('Index for instance types', top_candidate)
```
ditto

kgaikwad reviewed (changes requested):
@upadhyeammit, added few inline suggestions.

kgaikwad commented on ros/api/common/instance_descriptions.py:
```diff
@@ -0,0 +1,13 @@
+
+
+@cache.cached(timeout=0)
+def descriptions():
```
```suggestion
def instance_types_desc_dict():
```

kgaikwad commented on ros/api/v1/suggested_instance_types.py:
```diff
@@ -0,0 +1,75 @@
+    offset_value,
+    build_paginated_system_list_response
+)
+from ros.api.common.instance_descriptions import descriptions
```
```suggestion
from ros.api.common.instance_types_helper import descriptions
```

Details are for instance types and not for actual instances so in my opinion it would be good to rename file accordingly.

kgaikwad reviewed (changes requested):
Added two inline nitpick suggestions.

r14chandra reviewed (commented):
The sorting and filtering part looks good.

kgaikwad commented on ros/openapi/openapi.json:
```diff
@@ -295,6 +295,25 @@
                 }
             }
+        },
+        "/suggested_instance_types": {
```
This file needs to be updated with orderBy , orderHow & filter information.

kgaikwad commented on ros/openapi/openapi.json:
```diff
@@ -295,6 +295,25 @@
+        "/suggested_instance_types": {
+            "get": {
+                "summary": "List all suggested instance types for the system",
+                "description": "Lists the systems and their suggested instance type as per the Resource Optimization Rules",
```
```suggestion
                "description": "Lists suggested instance type as per the suggestions",
```

kgaikwad reviewed (commented):
Codewise rest changes look good to me :+1:

r14chandra commented on ros/lib/models.py:
```diff
@@ -18,6 +18,8 @@ class PerformanceProfile(db.Model):
     system_id = db.Column(db.Integer)
     psi_enabled = db.Column(db.Boolean)
+    top_candidate = db.Column(db.String(25), nullable=True)
+    top_candidate_price = db.Column(db.Float)
```
IIRC, the schema for `PerformanceProfile` and `PerformanceProfileHistory` is the same. To be consistent on this, we should add these fields to the history table as well.
@kgaikwad, let me know if I'm wrong.

upadhyeammit replied:
You are not wrong however we are not going to show the SuggestedInstaces from history table and hence it feels like information that we are not going to use.

r14chandra commented on ros/lib/utils.py:
```diff
@@ -151,15 +151,37 @@ def insert_performance_profiles(session, system_id, fields):
         session.flush()
 
 
+def redact_instance_type_and_price(fields):
```
(Based on my above comment) If we add those fields to the history table then we don't require this function and any changes to `insert_performance_profiles`.

upadhyeammit commented on ros/openapi/openapi.json:
```diff
@@ -935,6 +1005,30 @@
+                "name": "instance_type",
+                "in": "query",
+                "required": false,
+                "description": "Filter suggested instances by instance type",
```
Your suggestion makes sense!

r14chandra commented:
/retest

kgaikwad commented:
+ adding @yash2189 into loop - Could you give one round of testing for this PR and share observations?

yash2189 commented:
I have started progress on this. Will update with my observations upon completion

kgaikwad commented:
@upadhyeammit - could you perform rebase as it is having conflicts for hosts file?

upadhyeammit commented:
I have rebased this however there would be slight change in behavior ones we would get https://github.com/RedHatInsights/ros-backend/pull/364 merged!

upadhyeammit commented:
Rebased!

r14chandra commented:
> I have started progress on this. Will update with my observations upon completion

@yash2189, May I get any updates on this?

yash2189 commented:
Sorry @r14chandra I would have to check this again as I was working on other priority tasks. I will take a look at it soon

yash2189 commented:
/retest

upadhyeammit commented:
I have added the unit tests for multiple scenarios. In db fixtures I have added new methods to match system and performance_profile so in future if anyone requires that specific combo then we have it.

yash2189 commented:
Testing is completed on the PR. The feature is working as expected.
